File: src/libxpc/MacOSX/XPCSharedTypes/Sources/XPCSharedTypes/SyncInfo.swift
/*
 * Infomaniak kDrive - Desktop
 * Copyright (C) 2023-2025 Infomaniak Network SA
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import Foundation

protocol SyncInfoProtocol {
    var dbId: NSInteger { get }
    var driveDbId: NSInteger { get }
    var localPath: NSString { get }
    var targetPath: NSString { get }
    var targetNodeId: NSString { get }
    var supportVfs: Bool { get }
    var virtualFileMode: NSInteger { get }
    var navigationPaneClsid: NSString { get }

    init(dbId: NSInteger,
         driveDbId: NSInteger,
         localPath: NSString,
         targetPath: NSString,
         targetNodeId: NSString,
         supportVfs: Bool,
         virtualFileMode: NSInteger,
         navigationPaneClsid: NSString)
}

@objc public class SyncInfo: NSObject, NSSecureCoding, SyncInfoProtocol {
    public static let supportsSecureCoding = true

    public let dbId: NSInteger
    public let driveDbId: NSInteger
    public let localPath: NSString
    public let targetPath: NSString
    public let targetNodeId: NSString
    public let supportVfs: Bool
    public let virtualFileMode: NSInteger
    public let navigationPaneClsid: NSString

    public required init(dbId: NSInteger,
                         driveDbId: NSInteger,
                         localPath: NSString,
                         targetPath: NSString,
                         targetNodeId: NSString,
                         supportVfs: Bool,
                         virtualFileMode: NSInteger,
                         navigationPaneClsid: NSString) {
        self.dbId = dbId
        self.driveDbId = driveDbId
        self.localPath = localPath
        self.targetPath = targetPath
        self.targetNodeId = targetNodeId
        self.supportVfs = supportVfs
        self.virtualFileMode = virtualFileMode
        self.navigationPaneClsid = navigationPaneClsid
    }

    public required init?(coder: NSCoder) {
        guard let localPath = coder.decodeObject(forKey: "localPath") as? NSString,
              let targetPath = coder.decodeObject(forKey: "targetPath") as? NSString,
              let targetNodeId = coder.decodeObject(forKey: "targetNodeId") as? NSString,
              let navigationPaneClsid = coder.decodeObject(forKey: "navigationPaneClsid") as? NSString else {
            return nil
        }

        let dbId = coder.decodeInteger(forKey: "dbId")
        let driveDbId = coder.decodeInteger(forKey: "driveDbId")
        let virtualFileMode = coder.decodeInteger(forKey: "virtualFileMode")
        let supportVfs = coder.decodeBool(forKey: "supportVfs")

        self.dbId = dbId
        self.driveDbId = driveDbId
        self.localPath = localPath
        self.targetPath = targetPath
        self.targetNodeId = targetNodeId
        self.supportVfs = supportVfs
        self.virtualFileMode = virtualFileMode
        self.navigationPaneClsid = navigationPaneClsid
    }

    public func encode(with coder: NSCoder) {
        coder.encode(dbId, forKey: "dbId")
        coder.encode(driveDbId, forKey: "driveDbId")
        coder.encode(localPath, forKey: "localPath")
        coder.encode(targetPath, forKey: "targetPath")
        coder.encode(targetNodeId, forKey: "targetNodeId")
        coder.encode(supportVfs, forKey: "supportVfs")
        coder.encode(virtualFileMode, forKey: "virtualFileMode")
        coder.encode(navigationPaneClsid, forKey: "navigationPaneClsid")
    }
}

File: src/libcommon/info/syncinfo.cpp
/*
 * Infomaniak kDrive - Desktop
 * Copyright (C) 2023-2026 Infomaniak Network SA
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "syncinfo.h"
#include "utility/utility.h"

static const auto syncInfoDbId = "dbId";
static const auto syncInfoDriveDbId = "driveDbId";
static const auto syncInfoLocalPath = "localPath";
static const auto syncInfoTargetPath = "targetPath";
static const auto syncInfoTargetNodeId = "targetNodeId";
static const auto syncInfoSupportVfs = "supportVfs";
static const auto syncInfoVirtualFileMode = "virtualFileMode";
static const auto syncInfoNavigationPaneClsid = "navigationPaneClsid";

namespace KDC {

SyncInfo::SyncInfo(SyncDbId dbId, DriveDbId driveDbId, const QString &localPath, const QString &targetPath,
                   const QString &targetNodeId, bool supportVfs, VirtualFileMode virtualFileMode,
                   const QString &navigationPaneClsid) :
    _dbId(dbId),
    _driveDbId(driveDbId),
    _localPath(localPath),
    _targetPath(targetPath),
    _targetNodeId(targetNodeId),
    _supportVfs(supportVfs),
    _virtualFileMode(virtualFileMode),
    _navigationPaneClsid(navigationPaneClsid) {}

void SyncInfo::toDynamicStruct(Poco::DynamicStruct &dstruct) const {
    CommonUtility::writeValueToStruct(dstruct, syncInfoDbId, _dbId);
    CommonUtility::writeValueToStruct(dstruct, syncInfoDriveDbId, _driveDbId);
    CommonUtility::writeValueToStruct(dstruct, syncInfoLocalPath, CommonUtility::qStr2CommString(_localPath));
    CommonUtility::writeValueToStruct(dstruct, syncInfoTargetPath, CommonUtility::qStr2CommString(_targetPath));
    CommonUtility::writeValueToStruct(dstruct, syncInfoTargetNodeId, CommonUtility::qStr2CommString(_targetNodeId));
    CommonUtility::writeValueToStruct(dstruct, syncInfoSupportVfs, _supportVfs);
    CommonUtility::writeValueToStruct(dstruct, syncInfoVirtualFileMode, _virtualFileMode);
    CommonUtility::writeValueToStruct(dstruct, syncInfoNavigationPaneClsid, CommonUtility::qStr2CommString(_navigationPaneClsid));
}

void SyncInfo::fromDynamicStruct(const Poco::DynamicStruct &dstruct) {
    CommonUtility::readValueFromStruct(dstruct, syncInfoDbId, _dbId);
    CommonUtility::readValueFromStruct(dstruct, syncInfoDriveDbId, _driveDbId);

    CommString localPath;
    CommonUtility::readValueFromStruct(dstruct, syncInfoLocalPath, localPath);
    _localPath = CommonUtility::commString2QStr(localPath);

    CommString targetPath;
    CommonUtility::readValueFromStruct(dstruct, syncInfoTargetPath, targetPath);
    _targetPath = CommonUtility::commString2QStr(targetPath);

    CommString targetNodeId;
    CommonUtility::readValueFromStruct(dstruct, syncInfoTargetNodeId, targetNodeId);
    _targetNodeId = CommonUtility::commString2QStr(targetNodeId);

    CommonUtility::readValueFromStruct(dstruct, syncInfoSupportVfs, _supportVfs);
    CommonUtility::readValueFromStruct(dstruct, syncInfoVirtualFileMode, _virtualFileMode);

    CommString navigationPaneClsid;
    CommonUtility::readValueFromStruct(dstruct, syncInfoNavigationPaneClsid, navigationPaneClsid);
    _navigationPaneClsid = CommonUtility::commString2QStr(navigationPaneClsid);
}

QDataStream &operator>>(QDataStream &in, SyncInfo &info) {
    int dbId = 0; // must be int to work with QDataStreamd
    int driveDbId = 0; // must be int to work with QDataStreamd
    in >> dbId >> driveDbId >> info._localPath >> info._targetPath >> info._targetNodeId >> info._supportVfs >>
            info._virtualFileMode >> info._navigationPaneClsid;
    info.setDbId(dbId);
    info.setDriveDbId(driveDbId);
    return in;
}

QDataStream &operator<<(QDataStream &out, const SyncInfo &info) {
    out << toInt(info._dbId) << toInt(info._driveDbId) << info._localPath << info._targetPath << info._targetNodeId
        << info._supportVfs << info._virtualFileMode << info._navigationPaneClsid;
    return out;
}

QDataStream &operator<<(QDataStream &out, const QList<SyncInfo> &list) {
    int count = static_cast<int>(list.size());
    out << count;
    for (int i = 0; i < count; i++) {
        SyncInfo info = list[i];
        out << info;
    }
    return out;
}

QDataStream &operator>>(QDataStream &in, QList<SyncInfo> &list) {
    int count = 0;
    in >> count;
    for (int i = 0; i < count; i++) {
        SyncInfo info;
        in >> info;
        list.push_back(info);
    }
    return in;
}

} // namespace KDC
